text/0036-client-network-variables.md

```diff
@@ -0,0 +1,142 @@
+# Client Network Variables
+[feature]: #feature
+
+- Start Date: `2021-12-12`
+- RFC PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- SDK PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/0000)
+
+# Summary
+[summary]: #summary
+
+Today, on the server may write to `NetworkVariables`.  This RFC introduces a bifurcation where in addition to the existing (server) NetworkVariables, we now have ClientNetworkVariables, which clients may write to.
+
+# Motivation
+[motivation]: #motivation
+
+To enable users to more simply communicate to the server (especially input) without having to use RPCs, this feature is proposed.  It is an extension of the existing `NetworkVariables` differing mainly in that clients can write to them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Today, one adds a (server) `NetworkVariable` by declaring one in a `NetworkBehaviour` like so:
+
+```
+public class NetworkVariableTest : NetworkBehaviour
+{
+    public NetworkVariable<int> SomeSeverVar = new NetworkVariable<int>();
+
+    // ...
+
+    void Update()
+    {
+        if (NetworkManager.isServer)
+        {
+            SomeServerVar.Value = 123; // OK
+        }
+
+        if (NetworkManager.isClient)
+        {
+            // InvalidOperationException: Client can't write to NetworkVariables
+            // SomeServerVar.Value = 456; // ERROR
+        }
+
+    }
+}
+```
+
+Note, writing to the `SomeServerVar` in the server case succeeds (and is replicated to all clients), but fails with indicated exception in the comment.
+
+This RFC proposes a dedicated `ClientNetworkVariable` type that will work like this:
+
+```
+public class NetworkVariableTest : NetworkBehaviour
+{
+    public NetworkVariable<int> SomeSeverVar = new NetworkVariable<int>();
+    public ClientNetworkVariable<int> SomeClientVar = new ClientNetworkVariable<int>();
+
+    // ...
+
+    void Update()
+    {
+        if (NetworkManager.isServer)
+        {
+            SomeServerVar.Value = 123; // OK
+            SomeClientVar.Value = 456; // ERROR!
+        }
+
+        if (NetworkManager.isClient)
+        {
+            SomeServerVar.Value = 123; // ERROR!
+            SomeClientVar.Value = 456; // OK
+        }
+
+    }
+}
+```
+
+When the `SomeClientVar` code is executed...
+- The `SomeClientVar` for the client on which the code ran is immediately udpated
+- The client then communicates this variable update change to the server
+- The server then broadcasts this to the other clients
+- Note, unlike pre-1.0 versions of NGO when clients could write to network variables, the server can **not** write to client network variables.
```

Comment:
What about hosts? They are clients, they should be able to write to ClientNetvars too.
Any reasons we're not doing an Owner check and no client/server check? This way, a server that owns a netvar could write to the same netvar a client would if they were the owner.
With the proposed design, in boss room we'd have to have two sets of variables for each state and duplicate our gameplay state assignment everywhere :(

Reply:
Ok, good catch on host stuff, I need to double check the behavior.  The intention is the client who is the host shouldn't be special.